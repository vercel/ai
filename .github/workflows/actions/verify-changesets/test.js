import assert from 'node:assert';
import { mock, test } from 'node:test';

import { verifyChangesets } from './index.js';

test('happy path', async () => {
  const event = {
    pull_request: {
      labels: [],
    },
  };
  const env = {
    CHANGED_FILES: '.changeset/some-happy-path.md',
  };

  const readFile = mock.fn(async path => {
    return `---\nai: patch\n@ai-sdk/provider: patch\n---\n## Test changeset`;
  });

  await verifyChangesets(event, env, readFile);

  assert.strictEqual(readFile.mock.callCount(), 1);
  assert.deepStrictEqual(readFile.mock.calls[0].arguments, [
    '../../../../.changeset/some-happy-path.md',
    'utf-8',
  ]);
});

test('ignores .changeset/README.md', async () => {
  const event = {
    pull_request: {
      labels: [],
    },
  };
  const env = {
    CHANGED_FILES: '.changeset/README.md',
  };

  const readFile = mock.fn(() => {});

  await verifyChangesets(event, env, readFile);

  assert.strictEqual(readFile.mock.callCount(), 0);
});

test('invalid file - not a .changeset file', async () => {
  const event = {
    pull_request: {
      labels: [],
    },
  };
  const env = {
    CHANGED_FILES: '.changeset/not-a-changeset-file.txt',
  };

  const readFile = mock.fn(() => {});

  await assert.rejects(
    () => verifyChangesets(event, env, readFile),
    Object.assign(new Error('Invalid file - not a .changeset file'), {
      path: '.changeset/not-a-changeset-file.txt',
    }),
  );

  assert.strictEqual(readFile.mock.callCount(), 0);
});

test('invalid .changeset file - no frontmatter', async () => {
  const event = {
    pull_request: {
      labels: [],
    },
  };
  const env = {
    CHANGED_FILES: '.changeset/invalid-changeset-file.md',
  };

  const readFile = mock.fn(async path => {
    return 'frontmatter missing';
  });
  await assert.rejects(
    () => verifyChangesets(event, env, readFile),
    Object.assign(new Error('Invalid .changeset file - no frontmatter found'), {
      path: '.changeset/invalid-changeset-file.md',
      content: 'frontmatter missing',
    }),
  );
  assert.strictEqual(readFile.mock.callCount(), 1);
  assert.deepStrictEqual(readFile.mock.calls[0].arguments, [
    '../../../../.changeset/invalid-changeset-file.md',
    'utf-8',
  ]);
});

test('minor update', async () => {
  const event = {
    pull_request: {
      labels: [],
    },
  };
  const env = {
    CHANGED_FILES: '.changeset/patch-update.md .changeset/minor-update.md',
  };

  const readFile = mock.fn(async path => {
    if (path.endsWith('patch-update.md')) {
      return `---\nai: patch\n---\n## Test changeset`;
    }

    return `---\n@ai-sdk/provider: minor\n---\n## Test changeset`;
  });

  await assert.rejects(
    () => verifyChangesets(event, env, readFile),
    Object.assign(
      new Error(
        `Invalid .changeset file - invalid version bump (only "patch" is allowed, see https://sdk.vercel.ai/docs/migration-guides/versioning). To bypass, add one of the following labels: minor, major`,
      ),
      {
        path: '.changeset/minor-update.md',
        content: '---\n@ai-sdk/provider: minor\n---\n## Test changeset',
      },
    ),
  );

  assert.strictEqual(readFile.mock.callCount(), 2);
  assert.deepStrictEqual(readFile.mock.calls[0].arguments, [
    '../../../../.changeset/patch-update.md',
    'utf-8',
  ]);
  assert.deepStrictEqual(readFile.mock.calls[1].arguments, [
    '../../../../.changeset/minor-update.md',
    'utf-8',
  ]);
});

test('minor update - with "minor" label', async () => {
  const event = {
    pull_request: {
      labels: [
        {
          name: 'minor',
        },
      ],
    },
  };
  const env = {
    CHANGED_FILES: '.changeset/patch-update.md .changeset/minor-update.md',
  };

  const readFile = mock.fn(async path => {
    if (path.endsWith('patch-update.md')) {
      return `---\nai: patch\n---\n## Test changeset`;
    }

    return `---\n@ai-sdk/provider: minor\n---\n## Test changeset`;
  });

  const message = await verifyChangesets(event, env, readFile);
  assert.strictEqual(
    message,
    'Skipping changeset verification - "minor" label found',
  );
});

test('major update - with "major" label', async () => {
  const event = {
    pull_request: {
      labels: [
        {
          name: 'major',
        },
      ],
    },
  };
  const env = {
    CHANGED_FILES: '.changeset/patch-update.md .changeset/major-update.md',
  };

  const readFile = mock.fn(async path => {
    if (path.endsWith('patch-update.md')) {
      return `---\nai: patch\n---\n## Test changeset`;
    }

    return `---\n@ai-sdk/provider: major\n---\n## Test changeset`;
  });

  const message = await verifyChangesets(event, env, readFile);
  assert.strictEqual(
    message,
    'Skipping changeset verification - "major" label found',
  );
});
