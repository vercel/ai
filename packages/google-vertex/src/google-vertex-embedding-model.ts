import {
  EmbeddingModelV2,
  TooManyEmbeddingValuesForCallError,
} from '@ai-sdk/provider';
import {
  combineHeaders,
  createJsonResponseHandler,
  postJsonToApi,
  resolve,
  Resolvable,
} from '@ai-sdk/provider-utils';
import { z } from 'zod';
import { googleVertexFailedResponseHandler } from './google-vertex-error';
import {
  GoogleVertexEmbeddingModelId,
  GoogleVertexEmbeddingSettings,
} from './google-vertex-embedding-settings';
import { GoogleVertexConfig } from './google-vertex-config';

export class GoogleVertexEmbeddingModel implements EmbeddingModelV2<string> {
  readonly specificationVersion = 'v2';
  readonly modelId: GoogleVertexEmbeddingModelId;

  private readonly config: GoogleVertexConfig;
  private readonly settings: GoogleVertexEmbeddingSettings;

  get provider(): string {
    return this.config.provider;
  }

  get maxEmbeddingsPerCall(): number {
    return 2048;
  }

  get supportsParallelCalls(): boolean {
    return true;
  }

  constructor(
    modelId: GoogleVertexEmbeddingModelId,
    settings: GoogleVertexEmbeddingSettings,
    config: GoogleVertexConfig,
  ) {
    this.modelId = modelId;
    this.settings = settings;
    this.config = config;
  }

  async doEmbed({
    values,
    headers,
    abortSignal,
  }: Parameters<EmbeddingModelV2<string>['doEmbed']>[0]): Promise<
    Awaited<ReturnType<EmbeddingModelV2<string>['doEmbed']>>
  > {
    if (values.length > this.maxEmbeddingsPerCall) {
      throw new TooManyEmbeddingValuesForCallError({
        provider: this.provider,
        modelId: this.modelId,
        maxEmbeddingsPerCall: this.maxEmbeddingsPerCall,
        values,
      });
    }

    const mergedHeaders = combineHeaders(
      await resolve(this.config.headers),
      headers,
    );

    const url = `${this.config.baseURL}/models/${this.modelId}:predict`;
    const { responseHeaders, value: response } = await postJsonToApi({
      url,
      headers: mergedHeaders,
      body: {
        instances: values.map(value => ({ content: value })),
        parameters: {
          outputDimensionality: this.settings.outputDimensionality,
        },
      },
      failedResponseHandler: googleVertexFailedResponseHandler,
      successfulResponseHandler: createJsonResponseHandler(
        googleVertexTextEmbeddingResponseSchema,
      ),
      abortSignal,
      fetch: this.config.fetch,
    });

    return {
      embeddings: response.predictions.map(
        prediction => prediction.embeddings.values,
      ),
      usage: {
        tokens: response.predictions.reduce(
          (tokenCount, prediction) =>
            tokenCount + prediction.embeddings.statistics.token_count,
          0,
        ),
      },
      response: { headers: responseHeaders },
    };
  }
}

// minimal version of the schema, focussed on what is needed for the implementation
// this approach limits breakages when the API changes and increases efficiency
const googleVertexTextEmbeddingResponseSchema = z.object({
  predictions: z.array(
    z.object({
      embeddings: z.object({
        values: z.array(z.number()),
        statistics: z.object({
          token_count: z.number(),
        }),
      }),
    }),
  ),
});
