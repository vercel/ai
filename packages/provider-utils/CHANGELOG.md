# @ai-sdk/provider-utils

## 2.2.0

### Minor Changes

- 5bc638d: AI SDK 4.2

### Patch Changes

- Updated dependencies [5bc638d]
  - @ai-sdk/provider@1.1.0

## 2.1.15

### Patch Changes

- d0c4659: feat (provider-utils): parseProviderOptions function

## 2.1.14

### Patch Changes

- Updated dependencies [0bd5bc6]
  - @ai-sdk/provider@1.0.12

## 2.1.13

### Patch Changes

- Updated dependencies [2e1101a]
  - @ai-sdk/provider@1.0.11

## 2.1.12

### Patch Changes

- 1531959: feat (provider-utils): add readable-stream to unified test server

## 2.1.11

### Patch Changes

- Updated dependencies [e1d3d42]
  - @ai-sdk/provider@1.0.10

## 2.1.10

### Patch Changes

- Updated dependencies [ddf9740]
  - @ai-sdk/provider@1.0.9

## 2.1.9

### Patch Changes

- Updated dependencies [2761f06]
  - @ai-sdk/provider@1.0.8

## 2.1.8

### Patch Changes

- 2e898b4: chore (ai): move mockId test helper into provider utils

## 2.1.7

### Patch Changes

- 3ff4ef8: feat (provider-utils): export removeUndefinedEntries for working with e.g. headers

## 2.1.6

### Patch Changes

- Updated dependencies [d89c3b9]
  - @ai-sdk/provider@1.0.7

## 2.1.5

### Patch Changes

- 3a602ca: chore (core): rename CoreTool to Tool

## 2.1.4

### Patch Changes

- 066206e: feat (provider-utils): move delay to provider-utils from ai

## 2.1.3

### Patch Changes

- 39e5c1f: feat (provider-utils): add getFromApi and response handlers for binary responses and status-code errors

## 2.1.2

### Patch Changes

- ed012d2: feat (provider): add metadata extraction mechanism to openai-compatible providers
- Updated dependencies [3a58a2e]
  - @ai-sdk/provider@1.0.6

## 2.1.1

### Patch Changes

- e7a9ec9: feat (provider-utils): include raw value in json parse results
- Updated dependencies [0a699f1]
  - @ai-sdk/provider@1.0.5

## 2.1.0

### Minor Changes

- 62ba5ad: release: AI SDK 4.1

## 2.0.8

### Patch Changes

- 00114c5: feat: expose IDGenerator and createIdGenerator

## 2.0.7

### Patch Changes

- 90fb95a: chore (provider-utils): switch to unified test server
- e6dfef4: feat (provider/fireworks): Support add'l image models.
- 6636db6: feat (provider-utils): add unified test server

## 2.0.6

### Patch Changes

- 19a2ce7: feat (provider/fireworks): Add image model support.
- 6337688: feat: change image generation errors to warnings
- Updated dependencies [19a2ce7]
- Updated dependencies [6337688]
  - @ai-sdk/provider@1.0.4

## 2.0.5

### Patch Changes

- 5ed5e45: chore (config): Use ts-library.json tsconfig for no-UI libs.
- Updated dependencies [5ed5e45]
  - @ai-sdk/provider@1.0.3

## 2.0.4

### Patch Changes

- Updated dependencies [09a9cab]
  - @ai-sdk/provider@1.0.2

## 2.0.3

### Patch Changes

- 0984f0b: feat (provider-utils): Add resolvable type and utility routine.

## 2.0.2

### Patch Changes

- Updated dependencies [b446ae5]
  - @ai-sdk/provider@1.0.1

## 2.0.1

### Patch Changes

- c3ab5de: fix (provider-utils): downgrade nanoid and secure-json-parse (ESM compatibility)

## 2.0.0

### Major Changes

- b469a7e: chore: remove isXXXError methods
- b1da952: chore (provider-utils): remove convertStreamToArray
- 8426f55: chore (ai):increase id generator default size from 7 to 16.
- db46ce5: chore (provider-utils): remove isParseableJson export

### Patch Changes

- dce4158: chore (dependencies): update eventsource-parser to 3.0.0
- dce4158: chore (dependencies): update nanoid to 5.0.8
- Updated dependencies [b469a7e]
- Updated dependencies [c0ddc24]
  - @ai-sdk/provider@1.0.0

## 2.0.0-canary.3

### Major Changes

- 8426f55: chore (ai):increase id generator default size from 7 to 16.

## 2.0.0-canary.2

### Patch Changes

- dce4158: chore (dependencies): update eventsource-parser to 3.0.0
- dce4158: chore (dependencies): update nanoid to 5.0.8

## 2.0.0-canary.1

### Major Changes

- b1da952: chore (provider-utils): remove convertStreamToArray

## 2.0.0-canary.0

### Major Changes

- b469a7e: chore: remove isXXXError methods
- db46ce5: chore (provider-utils): remove isParseableJson export

### Patch Changes

- Updated dependencies [b469a7e]
- Updated dependencies [c0ddc24]
  - @ai-sdk/provider@1.0.0-canary.0

## 1.0.22

### Patch Changes

- aa98cdb: chore: more flexible dependency versioning
- 7b937c5: feat (provider-utils): improve id generator robustness
- 811a317: feat (ai/core): multi-part tool results (incl. images)
- Updated dependencies [aa98cdb]
- Updated dependencies [1486128]
- Updated dependencies [7b937c5]
- Updated dependencies [3b1b69a]
- Updated dependencies [811a317]
  - @ai-sdk/provider@0.0.26

## 1.0.21

### Patch Changes

- Updated dependencies [b9b0d7b]
  - @ai-sdk/provider@0.0.25

## 1.0.20

### Patch Changes

- Updated dependencies [d595d0d]
  - @ai-sdk/provider@0.0.24

## 1.0.19

### Patch Changes

- 273f696: fix (ai/provider-utils): expose size argument in generateId

## 1.0.18

### Patch Changes

- 03313cd: feat (ai): expose response id, response model, response timestamp in telemetry and api
- Updated dependencies [03313cd]
- Updated dependencies [3be7c1c]
  - @ai-sdk/provider@0.0.23

## 1.0.17

### Patch Changes

- Updated dependencies [26515cb]
  - @ai-sdk/provider@0.0.22

## 1.0.16

### Patch Changes

- 09f895f: feat (ai/core): no-schema output for generateObject / streamObject

## 1.0.15

### Patch Changes

- d67fa9c: feat (provider/amazon-bedrock): add support for session tokens

## 1.0.14

### Patch Changes

- Updated dependencies [f2c025e]
  - @ai-sdk/provider@0.0.21

## 1.0.13

### Patch Changes

- Updated dependencies [6ac355e]
  - @ai-sdk/provider@0.0.20

## 1.0.12

### Patch Changes

- dd712ac: fix: use FetchFunction type to prevent self-reference

## 1.0.11

### Patch Changes

- Updated dependencies [dd4a0f5]
  - @ai-sdk/provider@0.0.19

## 1.0.10

### Patch Changes

- 4bd27a9: chore (ai/provider): refactor type validation
- 845754b: fix (ai/provider): fix atob/btoa execution on cloudflare edge workers
- Updated dependencies [4bd27a9]
  - @ai-sdk/provider@0.0.18

## 1.0.9

### Patch Changes

- Updated dependencies [029af4c]
  - @ai-sdk/provider@0.0.17

## 1.0.8

### Patch Changes

- Updated dependencies [d58517b]
  - @ai-sdk/provider@0.0.16

## 1.0.7

### Patch Changes

- Updated dependencies [96aed25]
  - @ai-sdk/provider@0.0.15

## 1.0.6

### Patch Changes

- 9614584: fix (ai/core): use Symbol.for
- 0762a22: feat (ai/core): support zod transformers in generateObject & streamObject

## 1.0.5

### Patch Changes

- a8d1c9e9: feat (ai/core): parallel image download
- Updated dependencies [a8d1c9e9]
  - @ai-sdk/provider@0.0.14

## 1.0.4

### Patch Changes

- 4f88248f: feat (core): support json schema

## 1.0.3

### Patch Changes

- Updated dependencies [2b9da0f0]
- Updated dependencies [a5b58845]
- Updated dependencies [4aa8deb3]
- Updated dependencies [13b27ec6]
  - @ai-sdk/provider@0.0.13

## 1.0.2

### Patch Changes

- Updated dependencies [b7290943]
  - @ai-sdk/provider@0.0.12

## 1.0.1

### Patch Changes

- d481729f: fix (ai/provider-utils): generalize to Error (DomException not always available)

## 1.0.0

### Major Changes

- 5edc6110: feat (provider-utils): change getRequestHeader() test helper to return Record (breaking change)

### Patch Changes

- 5edc6110: feat (provider-utils): add combineHeaders helper
- Updated dependencies [5edc6110]
  - @ai-sdk/provider@0.0.11

## 0.0.16

### Patch Changes

- 02f6a088: feat (provider-utils): add convertArrayToAsyncIterable test helper

## 0.0.15

### Patch Changes

- 85712895: feat (@ai-sdk/provider-utils): add createJsonStreamResponseHandler
- 85712895: chore (@ai-sdk/provider-utils): move test helper to provider utils

## 0.0.14

### Patch Changes

- 7910ae84: feat (providers): support custom fetch implementations

## 0.0.13

### Patch Changes

- Updated dependencies [102ca22f]
  - @ai-sdk/provider@0.0.10

## 0.0.12

### Patch Changes

- 09295e2e: feat (@ai-sdk/provider-utils): add download helper
- 043a5de2: fix (provider-utils): rename to isParsableJson
- Updated dependencies [09295e2e]
  - @ai-sdk/provider@0.0.9

## 0.0.11

### Patch Changes

- Updated dependencies [f39c0dd2]
  - @ai-sdk/provider@0.0.8

## 0.0.10

### Patch Changes

- Updated dependencies [8e780288]
  - @ai-sdk/provider@0.0.7

## 0.0.9

### Patch Changes

- 6a50ac4: feat (provider-utils): add loadSetting and convertAsyncGeneratorToReadableStream helpers
- Updated dependencies [6a50ac4]
  - @ai-sdk/provider@0.0.6

## 0.0.8

### Patch Changes

- Updated dependencies [0f6bc4e]
  - @ai-sdk/provider@0.0.5

## 0.0.7

### Patch Changes

- Updated dependencies [325ca55]
  - @ai-sdk/provider@0.0.4

## 0.0.6

### Patch Changes

- 276f22b: fix (ai/provider): improve request error handling

## 0.0.5

### Patch Changes

- Updated dependencies [41d5736]
  - @ai-sdk/provider@0.0.3

## 0.0.4

### Patch Changes

- 56ef84a: ai/core: fix abort handling in transformation stream

## 0.0.3

### Patch Changes

- 25f3350: ai/core: add support for getting raw response headers.
- Updated dependencies [d6431ae]
- Updated dependencies [25f3350]
  - @ai-sdk/provider@0.0.2

## 0.0.2

### Patch Changes

- eb150a6: ai/core: remove scaling of setting values (breaking change). If you were using the temperature, frequency penalty, or presence penalty settings, you need to update the providers and adjust the setting values.
- Updated dependencies [eb150a6]
  - @ai-sdk/provider@0.0.1

## 0.0.1

### Patch Changes

- 7b8791d: Rename baseUrl to baseURL. Automatically remove trailing slashes.
