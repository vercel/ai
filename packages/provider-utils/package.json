{
  "name": "@ai-sdk/provider-utils",
  "version": "2.2.0",
  "license": "Apache-2.0",
  "sideEffects": false,
  "main": "./dist/index.js",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "files": [
    "dist/**/*",
    "test/dist/**/*",
    "CHANGELOG.md"
  ],
  "scripts": {
    "build": "tsup",
    "build:watch": "tsup --watch",
    "clean": "rm -rf dist && rm -rf test/dist",
    "lint": "eslint \"./**/*.ts*\"",
    "type-check": "tsc --noEmit",
    "prettier-check": "prettier --check \"./**/*.ts*\"",
    "test": "pnpm test:node && pnpm test:edge",
    "test:edge": "vitest --config vitest.edge.config.js --run",
    "test:node": "vitest --config vitest.node.config.js --run"
  },
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.mjs",
      "require": "./dist/index.js"
    },
    "./test": {
      "types": "./test/dist/index.d.ts",
      "import": "./test/dist/index.mjs",
      "module": "./test/dist/index.mjs",
      "require": "./test/dist/index.js"
    }
  },
  "dependencies": {
    "@ai-sdk/provider": "1.1.0",
    "eventsource-parser": "^3.0.0",
    "nanoid": "^3.3.8",
    "secure-json-parse": "^2.7.0"
  },
  "devDependencies": {
    "@types/node": "20.17.24",
    "@vercel/ai-tsconfig": "workspace:*",
    "msw": "2.7.0",
    "tsup": "^8",
    "typescript": "5.6.3",
    "zod": "3.23.8"
  },
  "peerDependencies": {
    "zod": "^3.23.8"
  },
  "engines": {
    "node": ">=18"
  },
  "publishConfig": {
    "access": "public"
  },
  "homepage": "https://sdk.vercel.ai/docs",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/vercel/ai.git"
  },
  "bugs": {
    "url": "https://github.com/vercel/ai/issues"
  },
  "keywords": [
    "ai"
  ]
}
