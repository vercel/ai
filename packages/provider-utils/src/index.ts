export * from './combine-headers';
export { convertAsyncIteratorToReadableStream } from './convert-async-iterator-to-readable-stream';
export * from './delay';
export * from './extract-response-headers';
export * from './fetch-function';
export { createIdGenerator, generateId } from './generate-id';
export * from './get-error-message';
export * from './get-from-api';
export * from './is-abort-error';
export * from './load-api-key';
export { loadOptionalSetting } from './load-optional-setting';
export { loadSetting } from './load-setting';
export * from './parse-json';
export * from './post-to-api';
export * from './remove-undefined-entries';
export * from './resolve';
export * from './response-handler';
export * from './uint8-utils';
export * from './validate-types';
export * from './validator';
export * from './without-trailing-slash';

export type { IDGenerator } from './generate-id';
export { parseProviderOptions } from './parse-provider-options';
export type { CoreToolCall, ToolCall } from './types/tool-call';
export type { CoreToolResult, ToolResult } from './types/tool-result';
