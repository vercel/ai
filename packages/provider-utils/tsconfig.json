{
  "extends": "./node_modules/@vercel/ai-tsconfig/ts-library.json",
  "compilerOptions": {
    "composite": true,
    "rootDir": "src",
    "outDir": "dist"
  },
  "exclude": [
    "dist",
    "build",
    "node_modules",
    "tsup.config.ts",
    "test.d.ts"
  ],
  "references": [
    {
      "path": "../provider"
    }
  ]
}
