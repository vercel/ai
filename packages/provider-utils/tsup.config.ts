import { defineConfig } from 'tsup';

export default defineConfig([
  {
    entry: ['src/index.ts'],
    format: ['cjs', 'esm'],
    dts: true,
    sourcemap: true,
    // Keep library target conservative for wide compatibility
    target: 'es2018',
    platform: 'node',
    define: {
      __PACKAGE_VERSION__: JSON.stringify(
        (await import('./package.json', { with: { type: 'json' } })).default
          .version,
      ),
    },
  },
  {
    entry: ['src/test/index.ts'],
    outDir: 'dist/test',
    format: ['cjs', 'esm'],
    dts: true,
    sourcemap: true,
    // Chai uses BigInt literals; ensure the target supports it and avoid bundling chai
    target: 'es2020',
    platform: 'node',
    external: [
      'chai',
      'vitest',
      'vitest/*',
      'msw',
      'msw/*',
      '@vitest/*',
      'vitest/dist/*',
      'vitest/dist/chunks/*',
      'vitest/dist/node/*',
      'vitest/dist/node/chunks/*',
    ],
    define: {
      __PACKAGE_VERSION__: JSON.stringify(
        (await import('./package.json', { with: { type: 'json' } })).default
          .version,
      ),
    },
  },
]);
