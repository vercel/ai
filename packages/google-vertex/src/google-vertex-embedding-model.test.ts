import {
  EmbeddingModelV1Embedding,
  TooManyEmbeddingValuesForCallError,
} from '@ai-sdk/provider';
import { createTestServer } from '@ai-sdk/provider-utils/test';
import { GoogleVertexEmbeddingModel } from './google-vertex-embedding-model';

const dummyEmbeddings = [
  [0.1, 0.2, 0.3],
  [0.4, 0.5, 0.6],
];
const testValues = ['test text one', 'test text two'];

const DEFAULT_URL =
  'https://us-central1-aiplatform.googleapis.com/v1/projects/test-project/locations/us-central1/publishers/google/models/textembedding-gecko@001:predict';

const CUSTOM_URL =
  'https://custom-endpoint.com/models/textembedding-gecko@001:predict';

const server = createTestServer({
  [DEFAULT_URL]: {},
  [CUSTOM_URL]: {},
});

describe('GoogleVertexEmbeddingModel', () => {
  const mockModelId = 'textembedding-gecko@001';
  const mockSettings = { outputDimensionality: 768 };
  const mockConfig = {
    provider: 'google-vertex',
    region: 'us-central1',
    project: 'test-project',
    headers: () => ({}),
    baseURL:
      'https://us-central1-aiplatform.googleapis.com/v1/projects/test-project/locations/us-central1/publishers/google',
  };

  const model = new GoogleVertexEmbeddingModel(
    mockModelId,
    mockSettings,
    mockConfig,
  );

  function prepareJsonResponse({
    embeddings = dummyEmbeddings,
    tokenCounts = [1, 1],
    headers,
  }: {
    embeddings?: EmbeddingModelV1Embedding[];
    tokenCounts?: number[];
    headers?: Record<string, string>;
  } = {}) {
    server.urls[DEFAULT_URL].response = {
      type: 'json-value',
      headers,
      body: {
        predictions: embeddings.map((values, i) => ({
          embeddings: {
            values,
            statistics: { token_count: tokenCounts[i] },
          },
        })),
      },
    };
  }

  it('should extract embeddings', async () => {
    prepareJsonResponse();

    const { embeddings } = await model.doEmbed({ values: testValues });

    expect(embeddings).toStrictEqual(dummyEmbeddings);
  });

  it('should expose the raw response headers', async () => {
    prepareJsonResponse({
      headers: {
        'test-header': 'test-value',
      },
    });

    const { rawResponse } = await model.doEmbed({ values: testValues });

    expect(rawResponse?.headers).toStrictEqual({
      // default headers:
      'content-length': '159',
      'content-type': 'application/json',
      // custom header
      'test-header': 'test-value',
    });
  });

  it('should extract usage', async () => {
    prepareJsonResponse({
      tokenCounts: [10, 15],
    });

    const { usage } = await model.doEmbed({ values: testValues });

    expect(usage).toStrictEqual({ tokens: 25 });
  });

  it('should pass the model parameters correctly', async () => {
    prepareJsonResponse();

    await model.doEmbed({ values: testValues });

    expect(await server.calls[0].requestBodyJson).toStrictEqual({
      instances: testValues.map(value => ({ content: value })),
      parameters: {
        outputDimensionality: mockSettings.outputDimensionality,
      },
    });
  });

  it('should pass headers correctly', async () => {
    prepareJsonResponse();

    const model = new GoogleVertexEmbeddingModel(mockModelId, mockSettings, {
      ...mockConfig,
      headers: () => ({
        'X-Custom-Header': 'custom-value',
      }),
    });

    await model.doEmbed({
      values: testValues,
      headers: {
        'X-Request-Header': 'request-value',
      },
    });

    expect(server.calls[0].requestHeaders).toStrictEqual({
      'content-type': 'application/json',
      'x-custom-header': 'custom-value',
      'x-request-header': 'request-value',
    });
  });

  it('should throw TooManyEmbeddingValuesForCallError when too many values provided', async () => {
    const tooManyValues = Array(2049).fill('test');

    await expect(model.doEmbed({ values: tooManyValues })).rejects.toThrow(
      TooManyEmbeddingValuesForCallError,
    );
  });

  it('should use custom baseURL when provided', async () => {
    server.urls[CUSTOM_URL].response = {
      type: 'json-value',
      body: {
        predictions: dummyEmbeddings.map(values => ({
          embeddings: {
            values,
            statistics: { token_count: 1 },
          },
        })),
      },
    };

    const modelWithCustomUrl = new GoogleVertexEmbeddingModel(
      'textembedding-gecko@001',
      { outputDimensionality: 768 },
      {
        headers: () => ({}),
        baseURL: 'https://custom-endpoint.com',
        provider: 'google-vertex',
      },
    );

    const response = await modelWithCustomUrl.doEmbed({
      values: testValues,
    });

    expect(response.embeddings).toStrictEqual(dummyEmbeddings);

    expect(server.calls[0].requestUrl).toBe(
      'https://custom-endpoint.com/models/textembedding-gecko@001:predict',
    );
  });

  it('should use custom fetch when provided and include proper request content', async () => {
    const customFetch = vi.fn().mockResolvedValue(
      new Response(
        JSON.stringify({
          predictions: dummyEmbeddings.map(values => ({
            embeddings: {
              values,
              statistics: { token_count: 1 },
            },
          })),
        }),
      ),
    );

    const modelWithCustomFetch = new GoogleVertexEmbeddingModel(
      'textembedding-gecko@001',
      { outputDimensionality: 768 },
      {
        headers: () => ({}),
        baseURL: 'https://custom-endpoint.com',
        provider: 'google-vertex',
        fetch: customFetch,
      },
    );

    const response = await modelWithCustomFetch.doEmbed({
      values: testValues,
    });

    expect(response.embeddings).toStrictEqual(dummyEmbeddings);

    expect(customFetch).toHaveBeenCalledWith(CUSTOM_URL, expect.any(Object));

    const [_, secondArgument] = customFetch.mock.calls[0];
    const requestBody = JSON.parse(secondArgument.body);

    expect(requestBody).toStrictEqual({
      instances: testValues.map(value => ({ content: value })),
      parameters: {
        outputDimensionality: 768,
      },
    });
  });
});
