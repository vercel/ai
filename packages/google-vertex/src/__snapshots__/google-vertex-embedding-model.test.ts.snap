// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GoogleVertexEmbeddingModel > should expose the raw response 1`] = `
{
  "body": {
    "predictions": [
      {
        "embeddings": {
          "statistics": {
            "token_count": 1,
          },
          "values": [
            0.1,
            0.2,
            0.3,
          ],
        },
      },
      {
        "embeddings": {
          "statistics": {
            "token_count": 1,
          },
          "values": [
            0.4,
            0.5,
            0.6,
          ],
        },
      },
    ],
  },
  "headers": {
    "content-length": "159",
    "content-type": "application/json",
    "test-header": "test-value",
  },
}
`;
